### PR TITLE
[CONTP-350] unify inclusion/exclusion logic for container images 

### PR DESCRIPTION
**Pull request by adel121**
### What does this PR do?

This PR unifies the inclusion/exclusion filtering logic for container images.

Documentation will be updated in [this](https://github.com/DataDog/documentation/pull/26733) PR.

### Motivation

There are several parts that should be explained here:
- What the documentation says is supported/not supported
- What is actually supported in practice
- How users are using the feature

These 3 parts are interconnected, and any change that is done needs to be handled with care in order not to break any user setup. 

#### Current Documentation:

Currently, [the documentation](https://docs.datadoghq.com/containers/guide/container-discovery-management/?tab=datadogoperator#environment-variables) indicates that container metrics and logs can be filtered for inclusion/exclusion by specifying regex for container name, namespace, and image name. 

The documentation doesn't clarify if image name can include the image tag (or image digest), or if it should strictly include the image name without the tag nor the digest.

But then, the documentation gives some examples:
- `image:^dockercloud/network-daemon$`
- `image:ubuntu`

None of the examples given in the documentation includes the image tag or image digest.

#### What is supported in practice:

In practice, we make no check on whether the regex for image name contains a tag or digest, so such filters are accepted during configuration of the agent.

We basically use filtering by image for 2 main purposes:
- filtering metrics 
- filtering logs

When checking if a log should be filtered out, we use `Image.RawName` field of the container entity in workloadmeta. 
The `Image.RawName` contains the full image name, including the tag or digest. ([example](https://github.com/DataDog/datadog-agent/blob/def18e3815c6cc67d8669408032961804821687a/comp/core/autodiscovery/listeners/kubelet.go#L122))

When checking if a metric should be filtered out, we use `Image.Name` field of the container entity in workloadmeta.
The `Image.Name` contains the image name excluding its tag or digest. ([example](https://github.com/DataDog/datadog-agent/blob/def18e3815c6cc67d8669408032961804821687a/pkg/collector/corechecks/containers/kubelet/provider/summary/provider.go#L217-L222))

We also have other use cases, but in general we don't impose any requirement on whether the image name that we match against the image filter regex includes should include (or may) the tag/digest or not.

Due to this discrepancy, users can configure filters that result in inconsistent filtering behaviour. Here is an example:

Suppose that a user configures `DD_CONTAINER_EXCLUDE: "image:^bar$"`

Based on the documentation, the user would expect exclusion of logs and metrics emitted by containers having the image `bar`, regardless of the image tag or digest.

However, since metric exclusion is matching `Image.Name` against the regex, and log exclusion is matching `Image.RawName` against the regex, only the metric exclusion will result in a match (`bar` matches `^bar$`, but `bar:latest` doesn't match ^bar$`).

A similar issue occurs if a user configures `DD_CONTAINER_EXCLUDE: "image:^bar:latest"`. In this case, logs will be excluded, but metrics won't.

Given this, we already have an inconsistency problem.

We had (at least one) support ticket regarding this problem (CONS-6500).

#### How users are configuring filters

Checking how users are configuring image filters, we notice that they actually use almost all possible ways of configuring image name regex filter.

Here are some examples:
- `image:********@sha256:e2feace0e0f852ffa3a3b9031[REDACTED]$`
- `image:906394416424.dkr.ecr.us-west-2.amazonaws.com/aws-for-fluent-bit:latest`
- `image:^gcr.io/datadoghq/agent:latest$`
- `image:mysql:8`

This means that some users might already have inconsistent filtering behaviour without them knowing about it.

A lot of users are including image tag in the image filter regex, and the documentation says nothing about it. It doesn't say if this is supported or not. In practice, it is supported for logs exclusion, but not for metrics exclusion.

### Summary

Based on what was explained above, it is difficult to drop support for including image tags and/or image digest for 2 main reasons:
1. It will break the filtering behaviour for most customers who already include image tag in their filters
2. Based statistics, customers seem to be interested in filtering by image and tag, so, supporting this provides value to the user.
3. In theory, it is more flexible to give the user the ability to filter by whatever they want.

In order to achieve consistent filtering support without breaking any user setup, we are doing the following:
- When building container filters, preprocess image filter regex string as follows:
   - If the regex string doesn't contain any tag or digest, and is of the form `name$` (e.g. `^nginx$` or `nginx$` ), convert the regex to `name(@sha256)?:.*`
   - Else, keep the regex as was set by the user.
- Update the `IsExcluded` method of container filter:
   - Currently, the method receives an argument named `containerImage`.
   - If the containerImage is non empty, and doesn't contain a colon (i.e. no tag, no digest), append a `:`.
- All components calling `IsExcluded` are expected to pass the maximum information that they have. For instance, if a component has the image name with the tag, it should include the tag when calling `IsExcluded`. (Specifically, the metric collectors should use `Image.RawName` instead of `Image.Name`).  

With this, no existing user setup will be broken, and the filtering logic will be consistent and uniform everywhere.

### Describe how you validated your changes

We need to validate filtering is working as expected.

For this we need to test several cases on a sample deployment.

We will use the following deployment as an example:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: dummy-test-app
  labels:
    run: my-test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: dummy-test-app
      run: my-test
  template:
    metadata:
      labels:
        app: dummy-test-app
        run: my-test
    spec:
      containers:
        - name: ubuntu
          image: ubuntu@sha256:98706f0f213dbd440021993a82d2f70451a73698315370ae8615cc468ac06624
          command: ["/bin/bash", "-c", "--"]
          args: ["while true; do sleep 30; echo hello-ubuntu; done;"]
        - name: nginx
          image: nginx:1.19.0
          command: ["/bin/bash", "-c", "--"]
          args: ["while true; do sleep 30; echo hello-nginx; done;"]
```

It contains 2 containers, the first one having an image with a digest, the second one having an image with a tag.

#### Case 1: Default Behaviour: No filtering at all

Deploy the agent with the following:

```
datadog:
  kubelet:
    tlsVerify: false
  apiKeyExistingSecret: datadog-secret
  appKeyExistingSecret: datadog-secret
  logs:
    enabled: true
    containerCollectAll: true
```

Navigate to the metrics and logs explorers and verify that metrics and logs are reported for both containers:

image

image

image

This is to ensure the default behaviour is still conserved.

#### Case 2: Filtering Out

Deploy the agent as follows:

```
datadog:
  kubelet:
    tlsVerify: false
  apiKeyExistingSecret: datadog-secret
  appKeyExistingSecret: datadog-secret
  logs:
    enabled: true
    containerCollectAll: true
  containerExclude: "image:^docker.io/library/ubuntu$ image:nginx:1.19.0$"
```

With this, metrics and logs should be excluded for both containers:

image

However, logs and metrics should still be visible for other containers:

image

### CODE REVIEW

File: releasenotes/notes/make_image_filters_consistent-14fd44a352723e50.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fixes consistency issue with container image filters.
+    Depending on the agent configuration, filters were sometimes behaving differently
```
Comment (drichards-87) [inline releasenotes/notes/make_image_filters_consistent-14fd44a352723e50.yaml:12]:
```suggestion
    Depending on the Agent configuration, filters were sometimes behaving differently
```

File: releasenotes/notes/make_image_filters_consistent-14fd44a352723e50.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    for metrics and logs. For example, an image filter that worked for excluding logs
+    didn't work when used to exclude metrics, and vice versa. 
+    The exclusion logic is now guaranteed to be consistent between metrics and logs.
```
Comment (drichards-87) [inline releasenotes/notes/make_image_filters_consistent-14fd44a352723e50.yaml:15]:
```suggestion
    The exclusion logic is now consistent between metrics and logs.
```

**Review by drichards-87 — APPROVED** [review]
Left some minor feedback from Docs and approved the PR.

**Review by daniel-taf — APPROVED** [review]

**Review by L3n41c — APPROVED** [review]

**Pull request merged**